File: exercises/exercises-mpi-p/commrank.py
#!/usr/bin/env python
################################################################
################################################################
####
#### This program file is part of the book and course
####   Parallel Computing for Science and Engineering
#### by Victor Eijkhout, copyright 2013-8
####
#### commrank.py : communicator rank and size
####
################################################################
################################################################

from mpi4py import MPI

comm = MPI.COMM_WORLD

####
#### Exercise part 1:
#### - use the Get_size and Get_rank methods of the COMM class
#### - let each processor print out a message like
####   "Hello from processor 7 out of 12"
####

##solution
nprocs = comm.Get_size()
procno = comm.Get_rank()
print("Hello from processor %d out of %d" % (procno,nprocs))
##solved

####
#### Exercise part 2:
#### - print out the number of processes from process zero
####

##solution
if procno==0:
    print("Running with %d processes" % nprocs)
##solved

File: exercises/exercises-mpl-mpl/commrank.cxx
/****************************************************************
 ****
 **** This program file is part of the book 
 **** `Parallel programming for Science and Engineering'
 **** by Victor Eijkhout, eijkhout@tacc.utexas.edu
 ****
 **** copyright Victor Eijkhout 2020-2026
 ****
 **** MPI Exercise for the use of Comm_rank/size
 ****
 ****************************************************************/

#include <iostream>
#include <format>
using namespace std;
#include <mpl/mpl.hpp>

int main() {

  const mpl::communicator &comm_world = mpl::environment::comm_world();
  
  // Exercise part 1:
  // -- Use the 
  //    `size' and `rank' methods
  // -- Let each processor print out a message like
  //    "Hello from processor 7 out of 12"
  //    reporting its number and the total number.
  int nprocs, procno;
/**** your code here ****/

  std::cout << std::format(
/**** your code here ****/
			   ) << '\n';

  // Exercise part 2:
  // -- let only processs zero print out
  //    "There are 16 processes"
  //    reporting only the total number
/**** your code here ****/
  
  return 0;
}
